src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs

```diff
@@ -0,0 +1,589 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonWriterHelper
+    {
+        public static readonly byte[] s_newLineUtf8 = Encoding.UTF8.GetBytes(Environment.NewLine);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryWriteIndentation(Span<byte> buffer, int indent, out int bytesWritten)
+        {
+            Debug.Assert(indent % 2 == 0);
+
+            if (buffer.Length >= indent)
+            {
+                if (indent < 8)
+                {
+                    int i = 0;
+                    while (i < indent)
+                    {
+                        buffer[i++] = JsonConstants.Space;
+                        buffer[i++] = JsonConstants.Space;
+                    }
+                }
+                else
+                {
+                    buffer.Slice(0, indent).Fill(JsonConstants.Space);
+                }
+                bytesWritten = indent;
+                return true;
+            }
+            else
+            {
+                int i = 0;
+                while (i < buffer.Length - 1)
+                {
+                    buffer[i++] = JsonConstants.Space;
+                    buffer[i++] = JsonConstants.Space;
+                }
+                bytesWritten = i;
+                return false;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ValidateProperty(ref ReadOnlySpan<byte> propertyName)
```

Comment:
There are a lot of places that pass Span as `ref`. It makes the code very hard to reason about because of you can never tell whether the methods change the argument. Could you please change the Spans to be passed as `in` where the value is not changed by the caller and passing by reference is just used as an optimization?

Reply:
> and passing by reference is just used as an optimization?

Is it actually a meaningful optimization?   We don't do it in pretty much any other place we pass around spans, and if it was meaningful, I'd expect it to be something the JIT should do instead of forcing the developer.

Reply:
> Is it actually a meaningful optimization? 

In certain cases, it improves performance by 10-20% (`in` or `ref` rather than pass by value)

Reply:
> In certain cases, it improves performance by 10-20% (in or ref rather than pass by value)

This call site makes a 10-20% difference?  You can prove me wrong, but I find that really hard to believe.  And if it's true, to me that suggests a codegen problem that should be fixed rather than contorting our code.

I could believe that there are a few methods where for one reason or another the extra word argument pushes it over some threshold in some optimization, maybe some argument that could have been passed in a register is now passed on the stack, or some such thing.  But I'm very skeptical that it then translates into benefiting from passing around all of these by ref or in.

I don't like this practice for a few reasons:
1) It causes complications when trying to do things like stackalloc spans due to compiler errors when passing around multiple ref structs, some by ref.  And in many cases, the compiler is flagging very real problems; as Jeremy highlighted, the hack to use unsafe to stackalloc a pointer and create a span from it in order to silence the compiler error actually resulted in leaking garbage pointers up the stack.
2) We do not believe customers should have to write code like this, and applying such patterns across a codebase (and one that we're distributing in a source package no less) will suggest to folks "this is how you write code with spans", when it's not actually what we recommend.
3) If it's really, truly necessary to get good performance, then that suggests we have much deeper rooted issues that should be addressed at their core rather than working around it in this way.

Reply:
I still believe it is in our collective best interest to pass small structs like `Span` by value everywhere, for the same reasons mentioned [a while back](https://github.com/dotnet/corefx/issues/27238#issuecomment-368077007).

Since then we have made improvements in the jitted codegen for passing structs with SysV ABI conventions (eg dotnet/coreclr#18358) that should start to show how the SysV convention can handle by-value small structs effectively. For that convention by-reference passing of small structs may be a pessimization.

So if we are going to measure perf to motivate passing small structs by-ref, we should measure on both Windows and Linux and try to strike the right balance.

Reply:
> I don't like this practice for a few reasons

Do you have the same opinion for using `in` instead of `ref` as well?

> I still believe it is in our collective best interest to pass small structs

Fair, will update it to pass by value.

> And if it's true, to me that suggests a codegen problem that should be fixed rather than contorting our code.

OK. To unblock this PR I will switch to pass by value everywhere and re-visit passing by `in` and its performance in the future.

Reply:
> Do you have the same opinion for using in instead of ref as well?

Yes, for the reasons other than (1) which shouldn't be as impactful.

Reply:
> I still believe it is in our collective best interest to pass small structs

**Follow up question:**
Should we pass by value even if the span is being passed around in a deep call graph?

For example, something like this:
```C#
WriteStringValue(ReadOnlySpan<char> value, bool suppressEscaping = false)
=> WriteStringSuppressFalse(in ReadOnlySpan<char> value)
=> WriteStringByOptions(in ReadOnlySpan<char> value)
=> WriteStringMinimized(in ReadOnlySpan<char> escapedValue)
=> WriteStringValue(in ReadOnlySpan<char> escapedValue, ref int idx)
```

Reply:
This is example of callstack where it would be ok to use some amount of AggresiveInlining to reduce its depth.

For example, WriteStringSuppressFalse has just a single callsite, so it would ok to inline it into the caller - either using AggresiveInlining; or by just not having the separate method at all and just inlining the code in manually.

Reply:
> so it would ok to inline it into the caller - either using AggresiveInlining; or by just not having the separate method at all and just inlining the code in manually.

Makes sense. I don't want to remove the separate method since it is easier to read what the code is doing when its split up into smaller methods. I will keep # of callsites in mind and mark methods as AI if they have a single reference (and I notice a performance improvement when doing so).

Reply:
https://github.com/dotnet/corefx/issues/34568

Reply:
In general, yes.

Whether `in` actually helps depends on a lot of factors: what each level does with the span, the relative frequencies of operations on the span, the impact of other code in each method, whether the intermediate layers get inlined, the target ABI, and so on. And the perf characteristics will change over time as the jit improves handling of structs in general.

If -- as your example implies -- you simply pass the span down to a single callee then `in` and by-value (implicit by-ref, on windows) will generate identical code on Windows x64 as the jit can recognize that in the apparent by-value case no copy is needed.